### PR TITLE
36 post donation

### LINKED ISSUE #36: POST /donation/id <Route>
* Protected Route <YES>
* Description:
  - [x] Get user id from params
  - [x] Get donation data (donation_type, quantity, ...) from request body
  - [x] Add donation into database with user id
  - [x] Test POST /api/donation/:id route
* Request body:
```js
{
  food: number,
  money: number,
  clothes: number,
  description: "string",
  location: "string",
  deliver_time: '' ,
}
```
* Validation:
```js
{
  food: number().min(1),
  clothes: number().min(1),
  money: number().min(1),
  description: string().min(1).max(255),
  location: string().min(1).max(255).required(),
  deliver_time: string().required(),
}
```

* Response:
  * Success (200):
    ```js
    { 
      message: 'Donation added successfully',
    }
    ```
  * Failed (400):
    ```jsx
    {
      message: Joi validation error message,
    };
    ```
    ```jsx
    {
      message: "Cannot add donation, campaign not exists",
    };
    ```
    ```jsx
    {
      message: "You should enter money, piece of clothes, number of meals",
    };
    ```

### PR DESCRIPTION
**Pull request by ghost**
closes #36

### CODE REVIEW

File: server/database/models/donation.ts [hunk 1]
```diff
@@ -14,6 +19,18 @@ const Donation = sequelize.define('donations', {
...
+    type: DataTypes.STRING,
+    allowNull: true,
+  },
+  deliver_time: {
+    type: DataTypes.DATE,
+    allowNull: false,
+  },
+  location: {
+    type: DataTypes.STRING,
+    allowNull: true,
```
Comment (Raghad-Mezied) [inline server/database/models/donation.ts:32]: AllowNull by default is true so no need to add them, and is it ok for the location to be null?

File: server/controllers/donations/addDonation.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+/* eslint-disable camelcase */
+import { NextFunction, Response } from 'express';
+import { Donation } from '../../database/models';
+import { CustomedError } from '../../utils';
+import { paramsSchema, donationSchema } from '../../utils/validation';
+
+const addDonation = async (req: any, res: Response, next: NextFunction) => {
+  const donorId: number = req.user.id;
```
Comment (Raghad-Mezied) [inline server/controllers/donations/addDonation.ts:8]: - you can destructure the id here and rename it, and add this line inside the try

File: server/controllers/donations/addDonation.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  try {
+    await paramsSchema.validateAsync(req.params);
+    const { id } = req.params;
+    await donationSchema.validateAsync(req.body);
+    const {
+      food, clothes, money, description, location, deliver_time,
+    } = req.body;
```
Comment (Raghad-Mezied) [inline server/controllers/donations/addDonation.ts:15]: Add all the declarations at the top and then all the validations

File: server/controllers/donations/addDonation.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    if (!food && !clothes && !money) {
+      next(
+        new CustomedError(
+          'You should enter money, piece of clothes, number of meals',
+          400,
+        ),
+      );
+    }
+    await Donation.create({
+      campaignId: id,
```
Comment (Raghad-Mezied) [inline server/controllers/donations/addDonation.ts:25]: Rename it camaignId from the beginning (in the route)

File: server/controllers/donations/addDonation.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      donorId,
+      food,
+      clothes,
+      money,
+      description,
+      location,
+      deliver_time,
+    });
+    res.json({ message: 'Donation added successfully' });
```
Comment (Raghad-Mezied) [inline server/controllers/donations/addDonation.ts:34]: You are adding a new item to the database so what should the status code be?

File: server/controllers/donations/addDonation.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  } catch (error) {
+    if (error.name === 'ValidationError') {
+      next(new CustomedError(error.details[0].message, 400));
+    } else if (error.name === 'SequelizeForeignKeyConstraintError') {
+      next(new CustomedError('Cannot add donation, campaign not exists', 400));
+    } else {
+      next(error);
```
Comment (Raghad-Mezied) [inline server/controllers/donations/addDonation.ts:41]: You can check before if there is a campaign with this id or not before adding

File: server/controllers/donations/addDonation.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
+/* eslint-disable camelcase */
```
Comment (Raghad-Mezied) [inline server/controllers/donations/addDonation.ts:1]: Fix this

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

File: server/database/models/donation.ts [hunk 8]
```diff
@@ -17,6 +22,18 @@ const Donation = sequelize.define('donations', {
     allowNull: false,
     defaultValue: 0,
   },
+  description: {
+    type: DataTypes.STRING,
+    allowNull: true,
```
Comment (Raghad-Mezied) [inline server/database/models/donation.ts:27]: Remove the `allowNull` for the description.

File: server/controllers/donations/addDonation.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+      },
+    } = req;
+    const campaign = await Campaign.findByPk(campaignId, {
+      raw: true,
+    });
+    if (!campaign) {
+      throw new CustomedError('Cannot add donation, campaign not exists', 400);
+    }
+    await paramsSchema.validateAsync({ id: campaignId });
+    await donationSchema.validateAsync(req.body);
```
Comment (Raghad-Mezied) [inline server/controllers/donations/addDonation.ts:26]: You are using the params before validating it?

File: .eslintrc.json [hunk 10]
```diff
@@ -25,7 +25,8 @@
             "ts": "never",
             "json": "never"
             }
-        ]       
+        ],
+        "camelcase": [0, {"properties": "never"}]
```
Comment (Raghad-Mezied) [inline .eslintrc.json:29]: Fix the problem, don't disable it or edit the rule.

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

**Review by Raghad-Mezied — APPROVED** [review]

File: server/controllers/donations/addDonation.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    if (!campaign) {
+      throw new CustomedError('Cannot add donation, campaign not exists', 400);
+    }
+
+    await donationSchema.validateAsync(req.body);
+
+    if (!food && !clothes && !money) {
+      next(
+        new CustomedError(
+          'You should enter money, piece of clothes, number of meals',
```
Comment (HaniOlwan) [inline server/controllers/donations/addDonation.ts:39]: i think this is better for a message what do you think ? "You should donate either money, clothes or meals"

**Review by haniolwan — APPROVED** [review]
Great job mohammed i like the tests

**Review by DinaAlghazali — APPROVED** [review]

**Review by farah2003 — APPROVED** [review]

**Pull request merged**